### ISSUE TITLE
Issues with Godot 4.3

### ISSUE THREAD

**Issue by HolonProduction** [bitbrain/pandora#196]
Godot version: 4.3-beta3

**Describe the bug**
Refer to godotengine/godot#94150 for further details

TL;DR:
The type handler code uses `Script.has_source_code` to test if a script is valid.
This is `false` for scripts with binary tokenization. `Script.can_instantiate` should be used.

The MRP from the orginal issue does work with this change, but fails when copying the newest version of pandora in there and making the change.

I found that when running the game, data.pandora gets saved in a corrupted way:
{"type": "undefined", "value": "<CompressedTexture#84384>"} where it should be {"type": "resource", "value": "res://godot.png"}. As stated this does not happen with the pandora version from the MRP, I don't know which version that is though.

**To Reproduce**
- Use the MRP from godotengine/godot#94150
- Follow the steps there to observe the issue with binary tokenization
- exchange `has_source_code` with `can_instantiate` to see that it fixes the issue
- copy the newest version of pandora to the addons folder
- exchange `has_source_code` with `can_instantiate` to see that it does not work anymore
- observe `data.pandora` to see that it get's corrupted when running the game

**Expected behavior**
- the texture should load when run through the editor and when exported with binary tokenization

**Screenshots**
- see godotengine/godot#94150

**Desktop (please complete the following information):**
 - OS: Fedora
 - Version 40

**Comment by shomykohai** [comment]
The MRP uses what used to be the latest commit on master, which in this case is 90f451445070a0bb114787c2971087fedb47c08f.

Can you test if applying the changes of https://github.com/bitbrain/pandora/pull/189 on the latest master commit, the issue gets fixed?

Edit: I've just looked upon my `data.pandora` and I encounter the data corruption problem too.
![broken types](https://github.com/user-attachments/assets/ea3cb784-162c-421e-afec-249edddb3991)

**Comment by shomykohai** [comment]
I was able to find where the problem comes from.
In `script_util.gd` it is possible to change `has_source_code` to `can_instantiate`, but on `type.gd` calling `can_instantiate` returns false.
![](https://github.com/user-attachments/assets/2ff35cbb-0cda-4f60-a42f-6e4fc52bd15a)

Apparently the scripts it tries to load are the one under `model/types`

**Comment by HolonProduction** [comment]
Just a guess, can't test right now: entity has `@tool` the type handlers not, so maybe `can_instantiate` returns false only in the Edtior for scripts without `@tool`. Could be verified by fixing the data file and running the project from the project manager without using the editor.

**Comment by shomykohai** [comment]
Indeed that does work, both on the ProjectManager and on exported builds (without a corrupted data.pandora).
I'd say this is something @bitbrain has to look through to decide on how to handle this.
I'd do something like:
```gdscript
if ScriptType != null:
  if Engine.is_editor_hint() and ScriptType.has_source_code():
			  return ScriptType.new()
  else:
    if ScriptType.can_instantiate():
			return ScriptType.new()
```
Not the best code for sure, but the workflow is that.

**Comment by bitbrain** [comment]
@shomykohai that code looks good. We should go with that.

**Comment by shomykohai** [comment]
Perfect, I'll make a PR for it.

**Comment by HolonProduction** [comment]
Why not make the type handlers tools though? If `can_instantiate` returns `false` it might just be a mistake or at least undefined behavior that it works correctly, and thus might arbitrary change in the future. (The type handlers are used in the editor, right?)

**Comment by shomykohai** [comment]
The function that gets called is a static function, that the editor uses to fetch the icons of each types.
When creating a property, they look to be wrapped around a PandoraProperty class, so I guess they're not meant to be tool classes.

Not knowing how fully Pandora works (I didn't look up the editor part that much) I can't say it for sure though.

Edit: just wondering, why does can_instantiate return false, assuming the script to be valid, when it does not have the `@tool` notation?
According to `gdscript.cpp`, it should only check for `valid`, right?

**Comment by HolonProduction** [comment]
Seems to also check for tool:
![image](https://github.com/user-attachments/assets/65a9b9c9-37b9-49ab-a9e5-c6f4e15aea62)
(No clue what this scripting enabled is :sweat_smile: )

I guess code that isn't marked as tool is not meant to run in editor which would include the initializer and constructor. Not sure what the intended behavior is with instantiating non tools from tool scripts. This question might better be asked in the contributors chat.

Back to pandora: I'd guess the editor has to use the non static `write_value` function when serializing. So IMO it would make sense for them to be tools, but @bitbrain would probably know best.
